File: flow_sdk/builtin/agentic_flow_run.py
"""AgenticFlowRun — one execution of an AgenticFlow (the run journal entity).

``execution_id`` (== this entity's id) stamps every event, delivery, and
spawned process of one activation, from the trigger/injection until the run
sinks. Parent = the AgenticFlow; spawned AgenticProcesses are attached as
children — so "everything this run did" is a plain child query.

The trace itself lives on disk (WORKFLOW_RUN precedent):
``<flow folder>/runs/<run-id>.jsonl`` — appended per event/phase by
FlowManager; this row is upserted only at run START and END so run traffic
never storms the DB or the indexer.
"""
from typing import ClassVar, Optional

from flow_sdk._compat import StrEnum
from flow_sdk.api.api_types.api_field import APIField
from flow_sdk.core import Entity
from flow_sdk.schema.types import EntityType


class RunStatus(StrEnum):
    RUNNING = "running"
    COMPLETE = "complete"    # sank: no pending deliveries, no active executions
    TRIPPED = "tripped"      # a loop budget refused further work
    FAILED = "failed"


class AgenticFlowRun(Entity):
    type: str = APIField(default=EntityType.AGENTIC_FLOW_RUN.value)
    name: str = APIField(default="")
    flow_id: str = APIField(default="", description="The AgenticFlow this run executed.")
    status: str = APIField(default=RunStatus.RUNNING.value)
    started_at: str = APIField(default="")
    ended_at: Optional[str] = APIField(None)
    # Cheap summary counters stamped at end (full trace = runs/<id>.jsonl).
    event_count: int = APIField(default=0)
    execution_count: int = APIField(default=0)
    error: Optional[str] = APIField(None, description="Budget trip / failure reason.")

    _api_visible: ClassVar[bool] = True

File: ts_sdk/src/entities/agentic-flow-run.ts
import { APIEntity, registerEntity } from '../APIEntity';
import { IEntity } from '../IEntity';

/**
 * AgenticFlowRun — one execution of an AgenticFlow (backend:
 * flow_sdk/builtin/agentic_flow_run.py). The row is start/end bookkeeping;
 * the full trace lives in the flow folder's runs/<id>.jsonl, served via
 * GET /api/v1/agentic-flows/<flow>/runs/<run>.
 */
export type AgenticFlowRunStatus = 'running' | 'complete' | 'tripped' | 'failed';

export interface IAgenticFlowRun extends IEntity {
  name?: string;
  flow_id?: string;
  status?: AgenticFlowRunStatus;
  started_at?: string;
  ended_at?: string;
  event_count?: number;
  execution_count?: number;
  error?: string;
}

@registerEntity
export class AgenticFlowRun extends APIEntity<AgenticFlowRun> implements IAgenticFlowRun {
  name?: string;
  flow_id?: string;
  status?: AgenticFlowRunStatus;
  started_at?: string;
  ended_at?: string;
  event_count?: number;
  execution_count?: number;
  error?: string;
  static type: string = 'agentic_flow_run';

  constructor(entity: Partial<IAgenticFlowRun> = {}) {
    super(entity);
    this.name = entity.name ?? '';
    this.flow_id = entity.flow_id ?? '';
    this.status = entity.status ?? 'running';
    this.started_at = entity.started_at;
    this.ended_at = entity.ended_at;
    this.event_count = entity.event_count ?? 0;
    this.execution_count = entity.execution_count ?? 0;
    this.error = entity.error;
  }
}
